compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/node/RootNodeOwner.kt

```diff
@@ -0,0 +1,498 @@
+
+        override fun calculateLocalPosition(positionInWindow: Offset): Offset = positionInWindow
+
+        private val endApplyChangesListeners = mutableVectorOf<(() -> Unit)?>()
+
+        override fun onEndApplyChanges() {
+            clearInvalidObservations()
+
+            // Listeners can add more items to the list and we want to ensure that they
+            // are executed after being added, so loop until the list is empty
+            while (endApplyChangesListeners.isNotEmpty()) {
+                val size = endApplyChangesListeners.size
+                for (i in 0 until size) {
+                    val listener = endApplyChangesListeners[i]
+                    // null out the item so that if the listener is re-added then we execute it again.
+                    endApplyChangesListeners[i] = null
+                    listener?.invoke()
+                }
+                // Remove all the items that were visited. Removing items shifts all items after
+                // to the front of the list, so removing in a chunk is cheaper than removing one-by-one
+                endApplyChangesListeners.removeRange(0, size)
+            }
+        }
+
+        override fun registerOnEndApplyChangesListener(listener: () -> Unit) {
+            if (listener !in endApplyChangesListeners) {
+                endApplyChangesListeners += listener
+            }
+        }
+
+        override fun registerOnLayoutCompletedListener(listener: Owner.OnLayoutCompletedListener) {
+            measureAndLayoutDelegate.registerOnLayoutCompletedListener(listener)
+            snapshotInvalidationTracker.requestLayout()
+        }
+    }
+
+    private inner class PlatformRootForTestImpl : PlatformRootForTest {
+        override val density get() = this@RootNodeOwner.density
+        override val textInputService get() = owner.textInputService
+        override val semanticsOwner get() = this@RootNodeOwner.semanticsOwner
+        override val visibleBounds: IntRect
+            get() {
+                val container = IntRect(IntOffset.Zero, platformContext.windowInfo.containerSize)
+                return bounds?.intersect(container) ?: container
+            }
+
+        override val hasPendingMeasureOrLayout: Boolean
+            get() = measureAndLayoutDelegate.hasPendingMeasureOrLayout
+
+        override fun measureAndLayoutForTest() {
+            owner.measureAndLayout(sendPointerUpdate = true)
+        }
+
+        /**
+         * Handles the input initiated by tests.
+         */
+        override fun sendPointerEvent(
+            eventType: PointerEventType,
+            position: Offset,
+            scrollDelta: Offset,
+            timeMillis: Long,
+            type: PointerType,
+            buttons: PointerButtons?,
+            keyboardModifiers: PointerKeyboardModifiers?,
+            nativeEvent: Any?,
+            button: PointerButton?
+        ) {
+            snapshotInvalidationTracker.onLayout()
+            measureAndLayout()
```

Comment:
We should move `measureAndLayout()` to `onPointerInput(event: PointerInputEvent)`. Otherwise, if we send an event to a scene with multiple owners, only one owner will be layouted. To proper handle an event, we should always layout the owner that receives the event.